sequencer/src/state_signature/relay_server.rs

```diff
@@ -0,0 +1,259 @@
+    threshold: U256,
+    /// Stake table
+    known_nodes: HashMap<StateVerKey, U256>,
+    /// Signatures packages for each block height
+    packages: HashMap<u64, StateSignaturesPackage>,
+
+    /// The latest state signatures package whose total weight exceeds the threshold
+    latest_available_package: Option<StateSignaturesPackage>,
+    /// The block height of the latest available state signature package
+    latest_block_height: Option<u64>,
+
+    /// A ordered queue of block heights, used for garbage collection.
+    queue: BTreeSet<u64>,
+
+    /// shutdown signal
+    shutdown: Option<OneShotReceiver<()>>,
+}
+
+impl StateRelayServerState {
+    pub fn new(threshold: U256) -> Self {
+        Self {
+            threshold,
+            ..Default::default()
+        }
+    }
+
+    pub fn with_shutdown_signal(mut self, shutdown_listener: Option<OneShotReceiver<()>>) -> Self {
+        if self.shutdown.is_some() {
+            panic!("A shutdown signal is already registered and can not be registered twice");
+        }
+        self.shutdown = shutdown_listener;
+        self
+    }
+}
+
+// TODO(Chengyu): move this `RwLock` inside `StateRelayServerState` so that when nodes are submitting
+//                signatures, it won't block the prover from fetching the available signatures.
+type State = RwLock<StateRelayServerState>;
+type Error = ServerError;
+
+pub trait StateRelayServerDataSource {
+    /// Get the latest available signatures package.
+    /// # Errors
+    /// Errors if there's no available signatures package.
+    fn get_latest_signature_package(&self) -> Result<StateSignaturesPackage, Error>;
+
+    /// Post a signature to the relay server
+    /// # Errors
+    /// Errors if the signature is invalid, already posted, or no longer needed.
+    fn post_signature(
+        &mut self,
+        key: StateVerKey,
+        state: LightClientState,
+        signature: StateSignature,
+    ) -> Result<(), Error>;
+}
+
+impl StateRelayServerDataSource for StateRelayServerState {
+    fn get_latest_signature_package(&self) -> Result<StateSignaturesPackage, Error> {
+        match &self.latest_available_package {
+            Some(package) => Ok(package.clone()),
+            None => Err(tide_disco::error::ServerError::catch_all(
+                StatusCode::NotFound,
+                "The light client state signatures are not ready.".to_owned(),
+            )),
+        }
+    }
+
+    fn post_signature(
+        &mut self,
+        key: StateVerKey,
+        state: LightClientState,
+        signature: StateSignature,
+    ) -> Result<(), Error> {
+        if (state.block_height as u64) <= self.latest_block_height.unwrap_or(0) {
+            return Err(tide_disco::error::ServerError::catch_all(
+                StatusCode::Locked,
+                "The posted signature is no longer needed.".to_owned(),
+            ));
+        }
+        let one = U256::one();
+        let weight = self.known_nodes.get(&key).unwrap_or(&one);
```

Comment:
Shouldn't the weight be 0 if the node is not in the stake table?

Reply:
`known_nodes` is empty here, so we can't do authentication. Once we have a stake table maintained, we should use the code commented below.

Reply:
oh lol